### PR TITLE
BinStat

### PR DESCRIPTION
**Pull request by kreisl**
New Statistics objects which can be created from Stats object. This should fix shortcomings of the old approach.
Drawing not yet implemented, but should be straight forward.

### CODE REVIEW

**Comment by kreisl** [comment]
Can you have a look if the class give us what we need? We could then also merge it with the other implementation of Stats Statistic and Resamples. Or keep them separate one class for gathering statistics and one for calculation of observables?

I used the arithmetics shown https://en.wikipedia.org/wiki/Algebra_of_random_variables and here https://link.springer.com/chapter/10.1007/978-1-4939-6572-4_4#Tab1 . We don't need to keep track of the moments to merge bins, I implemented the merging using the weighted average. This should also correctly merge the variance.
The formula is available here https://en.wikipedia.org/wiki/Pooled_variance#Sample-based_statistics .

**Comment by mam-mih-val** [comment]
@kreisl, I checked this on real data set. Rebin and projection seems to be working. But I've found, that 
1. After rebinnig and projection, errors in bins are zero. Should it be like this? 
2. If I do the same operation for instance calculation 3-sub resolution with stats and stat bins, errors from bootstrapping differs from each other

**Comment by kreisl** [comment]
1. Yes, I fixed it locally already. Will push changes after I resolve 2.
2. I think I accidentally calculated the variance of the sample not the variance of the sample mean, which is what we want. I need to fix the formulas. something like multiplying by (sum w)^2/ sum w^2

**Comment by kreisl** [comment]
Latest push should have fixed the issues seen previously, if I understood them correctly.

**Comment by mam-mih-val** [comment]
1D-container of stat bins works correctly, checked rebinnig and compared with initital container of stats. But seems to be one more bug appeared: rebinnig ang projection of multidimentional data container of stat bins refused to work. If I do rebinnig, it throws me that I have overlapping bins, even if I try to merge all them in one sigle bin through axis. And if I try to do projection, I get "nan" in means of bins, but error estimated normally.

**Comment by kreisl** [comment]
I added a 2-dimensional test case. I cannot reproduce the issue you are seeing. Are you just using this branch on top of master?

File: src/base/DataContainer.test.cpp [hunk 1]
```diff
@@ -67,11 +72,17 @@ TEST_F(DataContainerUnitTest, StatBinMerging) {
...
   EXPECT_NE(statbin_all[0].Mean(), statbin_part_1[0].Mean());
   EXPECT_NE(statbin_all[0].Mean(), statbin_part_2[0].Mean());
-  EXPECT_EQ(statbin_all[0].Mean(), manual_merge);
+  EXPECT_FLOAT_EQ(statbin_all[0].Mean(), manual_merge);
   EXPECT_FLOAT_EQ(statbin_all[0].Variance(), merged_statbin[0].Variance());
+  // bootstrap vs propagated errors
+  merged_statbin.SetErrors(Qn::StatBin::ErrorType::PROPAGATION);
+  auto prop_error = merged_statbin[0].StdDevOfMeanFromPropagation();
+  merged_statbin.SetErrors(Qn::StatBin::ErrorType::BOOTSTRAP);
+  auto bs_error = merged_statbin[0].StdDevOfMeanFromPropagation();
```
Comment (eugene274) [inline src/base/DataContainer.test.cpp]: It seems to be a copy-paste error

File: src/base/DataContainer.test.cpp [hunk 2]
```diff
@@ -81,6 +92,12 @@ TEST_F(DataContainerUnitTest, StatBinAddition) {
   auto man_sum_v = statbin_part_1[0].Variance() + statbin_part_2[0].Variance();
   EXPECT_FLOAT_EQ(sum[0].Mean(), man_sum);
   EXPECT_FLOAT_EQ(sum[0].Variance(), man_sum_v);
+  // bootstrap vs propagated errors
+  sum.SetErrors(Qn::StatBin::ErrorType::PROPAGATION);
+  auto prop_error = sum[0].StdDevOfMeanFromPropagation();
+  sum.SetErrors(Qn::StatBin::ErrorType::BOOTSTRAP);
+  auto bs_error = sum[0].StdDevOfMeanFromPropagation();
```
Comment (eugene274) [inline src/base/DataContainer.test.cpp]: also copy-paste error?

Comment (kreisl) [reply src/base/DataContainer.test.cpp]: yes i fixed them all

Comment (kreisl) [reply src/base/DataContainer.test.cpp]: thanks 👍

**Comment by mam-mih-val** [comment]
I'm sorry, that was my bad with overlapping bins, I set rebin range incorrectly. But still, I get "nan" mean and error of stat bins in case of projection multidimensional container of stat bins on single axis. I assume it might be because of neighboring empty bins in initial container. I mean, this bins are observable, which means we should take into account their weights, but somehow, attempt to merge two empty bins products nan.

File: include/base/QnTools/StatBin.hpp [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#ifndef QNTOOLS_STATBIN_H_
+
+#include <cmath>
+#include <vector>
+
```
Comment (eugene274) [inline include/base/QnTools/StatBin.hpp]: Missing headers: Rtypes.h, Stats.hpp

File: include/base/QnTools/StatBin.hpp [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+  /// \cond CLASSIMP
+ ClassDef(StatBin, 4);
+  /// \endcond
+};
+
+inline StatBin Merge(const StatBin &lhs, const StatBin &rhs) {
+  StatBin merged;
+  // mean and variance
+  merged.weight_ = lhs.Weight() + rhs.Weight();
+  merged.mean_ = (lhs.Mean() * lhs.Weight() + rhs.Mean() * rhs.Weight()) / merged.Weight();
```
Comment (eugene274) [inline include/base/QnTools/StatBin.hpp]: if `lhs.Mean()` or `rhs.Mean()` is _nan_, `merged.mean_` gets _nan_.

Test case:
* merge two empty StatBin-s
* merge result with filled StatBin

**Comment by eugene274** [comment]
@kreisl important offtop. When we discussed our Git workflow, you proposed keeping our development branches in private forks. I was rooting for having development branches in HeavyIonAnalysis/QnTools. You've submitted pull request. I wanted to add unit test reproducing Mikhail's issue. I cannot do that in you PR, because that is inside your private fork.

**Comment by kreisl** [comment]
@eugene274  as a maintainer you can push to other peoples pull requests. I think this is a good way of how to handle this procedure. For example: `git push git@github.com:user/repo local_branch_name:remote_branch_name`
should work

File: include/base/QnTools/StatBin.hpp (hunk 4 above)
Comment (kreisl) [reply include/base/QnTools/StatBin.hpp]: implemented a check based on empty weights

File: include/base/QnTools/StatBin.hpp (hunk 3 above)
Comment (kreisl) [reply include/base/QnTools/StatBin.hpp]: added

**Comment by eugene274** [comment]
> @eugene274 as a maintainer you can push to other peoples pull requests. I think this is a good way of how to handle this procedure. For example: `git push git@github.com:user/repo local_branch_name:remote_branch_name`
> should work

And it really works! Question closed. Thanks!

**Comment by kreisl** [comment]
Made another push cleaning up the classes a bit and fixing the bugs you have found. I notice the bs error seem to underestimate the statistical error. In some cases quite a bit. That's why the tests fail for some of the arithmetic operations. I am not sure if there is still a bug in the implementation or what deviation should be expected

**Comment by mam-mih-val** [comment]
@kreisl, I checked this out on real data sample again. It became better, means looks ok, but still in several bins there are "nan" errors. Checked error type. It's bootstrap. I think it's still the problem with merging empty bins.

**Comment by mam-mih-val** [comment]
@kreisl, propose: in StatBin::MergeBins() add special case for empty lhs and rhs in the same time, beacause situation improved, but somehow, some of bin errors still are nan.

File: src/base/StatBin.cpp [hunk 5]
```diff
@@ -48,9 +48,12 @@ StatBin Merge(const StatBin &lhs, const StatBin &rhs) {
...
   }
   // Bootstrap samples
-  if (lhs.sample_means_.size() != rhs.sample_means_.size()) {
+  if (lhs.sum_weight_ == 0.) {
     merged.sample_means_ = rhs.sample_means_;
     merged.sample_weights_ = rhs.sample_weights_;
+  } else if (rhs.sum_weight_ == 0.) {
+    merged.sample_means_ = lhs.sample_means_;
+    merged.sample_weights_ = lhs.sample_weights_;
   } else {
```
Comment (kreisl) [inline src/base/StatBin.cpp]: the only time when the division by sum of weights happens is when one of the bins which are being merged is not empty. so I am not sure why nan is still popping up?

Comment (kreisl) [reply src/base/StatBin.cpp]: > @kreisl, propose: in StatBin::MergeBins() add special case for empty lhs and rhs in the same time, beacause situation improved, but somehow, some of bin errors still are nan.

Can you send me the container before merging? I can have a look and try to debug.

Comment (mam-mih-val) [reply src/base/StatBin.cpp]: Okay, I'm sending you file with containers via email.

**Comment by kreisl** [comment]
There are Problems when merging bins with the variance from propagation. We need to fix the case where the weights are unequal 1. I am not sure which is the right formula. The variance from bootstrapping should not be affected by this problem.

**Comment by kreisl** [comment]
I made some new classes similar to what we discussed.
All of the accumulation of the statistics is happening using the StatCollect class.
It has two important members:
1. Statistics, which takes care of the summary statistics.
2. Bootstrap, which takes care of bootstrapping the standard error of the mean.
When we want to calculate an observable, or load the DataContainers containing the StatCollect from a files, we just Create a new one using `DataContainer<StatCalculate>(DataContainer<StatCollect> &)`. With this new container all of the operation to construct the variables can be performed.

**Pull request merged**